pkg/template/controller/templateinstance_finalizer.go

```diff
@@ -0,0 +1,279 @@
+	client            dynamic.DynamicInterface
+	config            *rest.Config
+	jsonConfig        *rest.Config
+	templateClient    templateclient.Interface
+
+	// FIXME: Remove then cient when the build configs are able to report the
+	//				status of the last build.
+	buildClient buildclient.Interface
+
+	kc kclientsetinternal.Interface
+
+	lister   templatelister.TemplateInstanceLister
+	informer cache.SharedIndexInformer
+
+	queue workqueue.RateLimitingInterface
+
+	readinessLimiter workqueue.RateLimiter
+
+	clock clock.Clock
+}
+
+// NewTemplateInstanceFinalizerController returns a new TemplateInstanceFinalizerController.
+func NewTemplateInstanceFinalizerController(dynamicRestMapper *discovery.DeferredDiscoveryRESTMapper, config *rest.Config, kc kclientsetinternal.Interface, buildClient buildclient.Interface, templateClient templateclient.Interface, informer internalversion.TemplateInstanceInformer) *TemplateInstanceFinalizerController {
+	c := &TemplateInstanceFinalizerController{
+		restmapper:        restutil.DefaultMultiRESTMapper(),
+		dynamicRestMapper: dynamicRestMapper,
+		config:            config,
+		kc:                kc,
+		templateClient:    templateClient,
+		buildClient:       buildClient,
+		lister:            informer.Lister(),
+		informer:          informer.Informer(),
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "openshift_template_instance_controller"),
+		readinessLimiter:  workqueue.NewItemFastSlowRateLimiter(5*time.Second, 20*time.Second, 200),
+		clock:             clock.RealClock{},
+	}
+
+	c.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(c.config)
+	c.jsonConfig.ContentConfig = dynamic.ContentConfig()
+	var err error
+	c.client, err = dynamic.NewForConfig(c.jsonConfig)
+	if err != nil {
+		glog.Errorf("failure creating dynamic client: %v", err)
+		return nil
+	}
+
+	return c
+}
+
+// getTemplateInstance returns the TemplateInstance from the shared informer,
+// given its key (dequeued from c.queue).
+func (c *TemplateInstanceFinalizerController) getTemplateInstance(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.lister.TemplateInstances(namespace).Get(name)
+}
+
+// sync is the actual controller worker function.
+func (c *TemplateInstanceFinalizerController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	needsFinalizing := false
```

Comment:
check deletiontimestamp again.  If some cleared a finalizer, then manually deleted, then recreated with the same name, you'd end up here again and you should skip action.